DESIGN.md

```diff
@@ -0,0 +1,756 @@
+large number of open connections, this can be problematic. The io-uring API
+supports registering buffer pools with the ring and configuring read operations
+to use the buffer pool instead of a dedicated per-operation buffer. When the
+socket receives data, the kernel checks out a buffer from the pool and returns
+it to the caller. After reading the data, the caller returns the buffer to the
+kernel. If no buffers are available, the operation pauses until the process
+returns a buffer to the kernel, at which time the operation completes.
+
+```rust
+// ...
+let buf = file.read_at_prepared(0, 1024, DefaultPool).await?;
+
+// Write some data back to the file.
+let BufResult(res, buf) = file.write_at(1024, buf.slice(0..512)).await;
+res?;
+
+// Dropping the buffer lets the kernel know the buffer may
+// be reused.
+drop(buf);
+```
+
+The runtime supports initializing multiple pools containing buffers of different
+sizes. When submitting a read operation, the caller specifies the pool from
+which to draw a buffer.
+
+```rust
+// Allocate a buffer pool
+let my_pool = BufferPool::builder()
+    .buffer_size(16_384)
+    .num_buffers(256)
+    .build();
+
+// Create the runtime
+let mut rt = tokio_uring::runtime::Runtime::new()?;
+
+// Provide the buffer pool to the kernel. This passes
+// ownership of the pool to the kernel.
+let pool_token = rt.provide_buffers(MY_POOL, 256, 4 * 4096)?;
+    
+rt.block_on(async {
+    // ...
+    let buf = file.read_at_prepared(0, 1024, pool_token).await?;
+});
+```
+
+## Buffer management
+
+Buffers passed to read and write operations must remain alive and pinned to a
+memory location while operations are in-flight, ruling out `&mut [u8]` as an
+option. Additionally, read and write operations may reference buffers using a
+pointer or, when the buffer is part of a pre-registered buffer pool, using a
+numeric buffer identifier. 
+
+The tokio-uring crate will provide its buffer type, `IoBuf`, to use when reading
+and writing.
+
+```rust
+pub struct IoBuf {
+    kind: Kind,
+}
+
+enum Kind {
+    /// A vector-backed buffer
+    Vec(Vec<u8>),
+
+    /// Buffer pool backed buffer. The pool is managed by io-uring.
+    Provided(ProvidedBuf),
+}
+```
+
+Internally, the IoBuf type is either backed by individually heap-allocated
+memory or a buffer pool entry. Acquiring an `IoBuf` is done via
+`IoBuf::with_capacity` or checking-out an entry from a pool. 
+
+```rust
+// Individually heap-allocated
+let my_buf = IoBuf::with_capacity(4096);
+
+// Checked-out from a pool
+match my_buffer_pool.checkout() {
+    Ok(io_buf) => ...,
+    Err(e) => panic!("buffer pool empty"),
+}
+```
+
+On drop, if the `IoBuf` is a buffer pool member, it is checked back in. If the
+kernel initially checked out the buffer as part of a read operation, an io-uring
+operation is issued to return it. Submitting the io-uring operation requires the
+buffer to remain on the same thread that checked it out and is enforced by
+making the `IoBuf` type `!Send`. 
```

Comment:
Does this also mean that the pools themselves are `!Send`? I suspect not, but it wasn't clear from this paragraph.

Reply:
They are `!Send` as they are registered w/ a specific uring.

Reply:
It'd be worthwhile to call that out explicitly in the doc I think.